pkg/controller/registryinfra/registryinfra_controller.go

```diff
@@ -0,0 +1,795 @@
+		logger.Errorf("retrying registry infra")
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logger.Infof("dropping cluster out of the queue: %v", err)
+	c.queue.Forget(key)
+}
+
+func (c *Controller) syncCluster(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing registry infra")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing registry infra")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	capiCluster, err := c.clusterLister.Clusters(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Debug("cluster has been deleted")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	cluster, err := controller.ConvertToCombinedCluster(capiCluster)
+	if err != nil {
+		return fmt.Errorf("failure converting to combined cluster: %v", err)
+	}
+
+	//clusterProviderStatus, err := controller.ClusterProviderStatusFromCluster(capiCluster)
+	_, err = controller.ClusterProviderStatusFromCluster(capiCluster)
+	if err != nil {
+		return fmt.Errorf("error retrieving clusterproviderstatus: %v", err)
+	}
+
+	// check if the cluster is being deleted
+	if capiCluster.DeletionTimestamp != nil {
+		//if true {
+		c.logger.Debugf("cleaning up registry infrastructure pieces for cluster %v", capiCluster.Name)
+
+		err = c.syncRegistryInfra(cluster, false /* deprovision infra */)
+		if err != nil {
+			return fmt.Errorf("error deleting registry infra: %v", err)
+		}
+		err = c.deleteFinalizer(capiCluster)
+		return err
+	}
+
+	if !hasRegistryInfraFinalizer(capiCluster) {
+		c.logger.Debugf("adding registryinfra finalizer to cluster %v", capiCluster.Name)
+		return c.addFinalizer(capiCluster)
+	}
+
+	if true {
+		//if !clusterProviderStatus.RegistryInfraInstalled {
+		c.logger.Debugf("creating registry infrastructure pieces for cluster %v", capiCluster.Name)
+
+		err = c.syncRegistryInfra(cluster, true /* create infra */)
+		if err != nil {
+			c.updateClusterStatus(capiCluster, false)
+			return fmt.Errorf("error setting up infrastructure for registry: %v", err)
+		}
+
+		err = c.updateClusterStatus(capiCluster, true)
+		if err != nil {
+			return fmt.Errorf("error updating cluster status: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) deprovisionS3Bucket(bucketName string, awsSession *session.Session) error {
+
+	s3Svc := s3.New(awsSession)
+
+	s3Bucket, err := getBucketByName(bucketName, s3Svc)
+	if err != nil {
+		return fmt.Errorf("error searching for existing bucket: %v", err)
+	}
+
+	if s3Bucket != nil {
+		_, err = s3Svc.DeleteBucket(&s3.DeleteBucketInput{
+			Bucket: aws.String(bucketName),
+		})
+		if err != nil {
+			return fmt.Errorf("error deleting bucket: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// create S3 bucket (if necessary) for cluster's registry
```

Comment:
Nit: godoc syntax is off, "setupS3Bucket creates the S3 bucket if necessary...".